llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp

```diff
@@ -605,12 +606,197 @@ class SIGfx12CacheControl : public SIGfx11CacheControl {
                                       bool IsNonTemporal) const override;
 };
 
+class SIPreciseMemorySupport {
```

Comment:
Why does it need to be a separate class hierarchy?
It could just be part of CacheControl, and the functions can be named `handlePreciseMemoryAtomic/NonAtomic` ?
That would avoid a lot of boilerplate.

Reply:
I don't have strong objection to merging with CacheControl, but would like to get Tony's opinion. @t-tye Tony, what do you think?

Reply:
My initial thought had been that this would be part of the existing cache control functions. It seems it is the same kind of waitcnt as needs to be inserted after a store release. That also requires the right waitcnt to be generated according to the kind of memory instruction.

Reply:
Ok, will merge with CacheControl.

Reply:
Merged with SICacheControl.